TaxHelpers#rates_for_item now respects the validity period of tax rates

jugglinghobo opened the pull request:
**Description**
When calculating the tax for line items whose tax is included in the price, solidus uses TaxRates in the calculation even though they are invalid according to their start_date and/or end_date. This results in incorrect tax calculations if there are multiple tax rates for the same tax category and only some of them are currently valid.

This PR ensures that `TaxHelpers#rates_for_item` only returns tax rates that are currently valid. Before, it would return any tax rate that matched the items' tax category.

This addresses https://github.com/solidusio/solidus/issues/3766

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [ ] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

jugglinghobo commented:
As suggested in https://github.com/solidusio/solidus/issues/3766, I have opened a PR and moved the change to the `TaxHelpers#rates_for_item` method.

I am currently still unsure about two things:

* Maybe it would make sense to add an `.active` scope to `TaxRate`. If so, would you like me to do this in this PR directly?
* I didn't create a new spec file for the `TaxHelpers` module and instead added a test to the `DefaultTax` calculator which is one of two places affected by the change. Would you like me to create a spec file for the `TaxHelpers` module? If so, where should it be (since `TaxHelpers` is in `models/` for some reason...?).

I'm very open for suggestions and improvements.

aldesantis commented:
@jugglinghobo thanks for this! Answers to your concerns below:

> * Maybe it would make sense to add an `.active` scope to `TaxRate`. If so, would you like me to do this in this PR directly?

Yes, I think it would be useful. 🙂  Feel free to do it in this PR!

> * I didn't create a new spec file for the `TaxHelpers` module and instead added a test to the `DefaultTax` calculator which is one of two places affected by the change. Would you like me to create a spec file for the `TaxHelpers` module? If so, where should it be (since `TaxHelpers` is in `models/` for some reason...?).

Yes, I think the test for this should be on the `TaxHelpers` module itself. You can put it in `spec/models/spree/tax/tax_helpers_spec.rb`. Testing modules can be a bit tricky (you'd usually create a "dummy" test class and include the module there, then test the method somehow), so let me know if you need any help.

mamhoff commented:
@aldesantis @jugglinghobo I think this is actually a pretty important fix functionality-wise. I appreciate wanting to have very good specs that respect all the things, and maybe also a scope, but can we make issues for that and get this in? I'm seeing clients write workarounds in their exporter code for tax advisors because this doesn't work. I'd also be happy to take this over if necessary. (cc @tvdeyen @mntmn).

jugglinghobo commented:
@mamhoff @aldesantis I'm sorry this fell off my radar. If you can wait, I'll try to finish this tonight or tomorrow. If you don't hear back until sunday or if you need a solution quicker, feel free to take over, and if you leave out anything (e.g. scope), feel free to assign an issue to me or whatever you'd like.

Does that work for you, or would you like me to do anything else? This is my first solidus / open source contribution.

aldesantis commented:
@jugglinghobo even if this is merged today, it still wouldn't be available until the next Solidus release, so tonight/tomorrow works perfectly. Thanks for your help on this one!

mamhoff commented:
I just didn't want this to sit for another couple of weeks. If you're on it, great! :+1:

jugglinghobo commented:
@aldesantis 

As discussed I've added a spec for `TaxHelpers#rates_for_item`. IMO this is all that's required to fix the bug.
I also added the `TaxRate.active` scope (identical to the one in `Promotion`), but it is unused at the moment.

Please suggest improvements and advise on how to proceed. I didn't squash yet because I thought maybe you'd like to add the scope in a separate PR.

PS: AFAICT there are no specs for `Promotion.active`, if the new specs are accepted then these could be reused there.

aldesantis commented on core/spec/models/spree/tax/tax_helpers_spec.rb:
```diff
@@ -0,0 +1,46 @@
+
+require 'rails_helper'
+
+class DummyClass
+  include Spree::Tax::TaxHelpers
+
+  def valid_rates(item)
+    rates_for_item(item)
+  end
+end
```
Instead of defining the class outside of the tests, can you do with `before` and `stub_const`?
```ruby
before do
  stub_const('DummyClass', (Class.new do
    include Spree::Tax::TaxHelpers

    def valid_rates(item)
      rates_for_item(item)
    end
  end))
end
```
The reason is that a class defined with the `class` keyword will leak between tests, while one defined with `stub_const` will be unloaded as soon as the test has finished executing.

jugglinghobo replied:
Thanks for the feedback, I've pushed the requested changes.

aldesantis reviewed (changes requested):
@jugglinghobo thanks, the tests are okay except for one minor thing. After that, GTM!

aldesantis reviewed (approved):
Looks great, thanks so much! 🙏

mamhoff reviewed (approved):
Nicely done!

kennyadsl reviewed (approved):
Thanks!

jugglinghobo commented:
Thank you guys!
